### PR TITLE
feat(fms): cFMS 1.5

### PR DESCRIPTION
**Pull request by Benjozork**
Credit @professoralex13 @tracernz for the help.

## Summary of Changes

* This PR introduces support for the following FMS features
  * LNAV
    * Holding patterns
    * Turn direction constraints on non-TF legs
    * Overfly restriction support
    * ARINC424 leg types
      * AF, CA, CI, CR, CF, DF, HF, HM, HA legs
    * Turn prediction types
      * Path capture
      * Course capture
      * Direct to fix turn
      * Holding pattern entry turn   
  * ND
    * Removed flight plan loading from `localStorage`
    * Corrected active waypoint ETA display
    * Added `EfisVectors` systems with optimized transmit task queue + support for **future** display of OFFSET, SECONDARY, SECONDARY DASHED, MISSED APPROACH, ALTERNATE and EOSID flight paths.

## Known limitations

- CA leg terminations are sometimes in the wrong place, and do not adapt to V/S.
- Some path captures will be incorrectly drawn. This will not affect guidance.
- INTCPT calculation can be off on large distances.
- Course captures do not adapt to PPOS when nextLeg is active. This can cause the final path to be off to the side.

## Screenshots (if necessary)

**LSZH DEGES3W Before**

image

**LSZH DEGES3W After**

image

**Overfly support on TF legs**

image

**Direct To Fix turns**

image

**Honeywell path captures**

image

## References

An absolutely enormous amount of them and a year+ of conception.

## Additional context

Discord username (if different from GitHub): someperson#4953

## Testing instructions

PHASE 1 - General LNAV and path generation

* Make sure to test loading various SIDs and STARs and that their drawing is reasonable
  * Test departures with turn direction constraints
* Make sure to test flying various SIDs and STARs and that the aircraft flies it correctly
  * Keep in mind that path capture turns (sharp lines joining the next leg with a 45 degree angle) will not be followed exactly - this is accurate to the IRL Honeywell FMS 
  * Check if turns change shape / size while flying them (they should not)
* Make sure sequencing and guidance work correctly (no turning the wrong way, minimal overshooting / undershooting)
  * Make sure VM (MANUAL) legs do not sequence and do not show a distance or ETA on the ND but do show a bearing
  * Make sure regular legs show a distance, bearing and ETA on the NDD
* Make sure ND does not crash and that there is no unacceptable delay in path drawing 

PHASE 2 - Holding patterns, entries and exits (not available yet)

TBD

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

File: src/fmgc/src/efis/EfisCommon.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        if (range <= 80) {
+            return [56, 56, 56];
+        }
+        if (range <= 160) {
+            return [112, 112, 112];
+        }
+        return [224, 224, 224];
+    default:
+        return [0, 0, 0];
+    }
```
Comment (davidwalschots) [inline src/fmgc/src/efis/EfisCommon.ts:76]: Random suggestion

```suggestion
    case Mode.PLAN:
        if (range <= 10) {
            return Array(3).fill(7);
        }
        if (range <= 20) {
            return Array(3).fill(14);
        }
        if (range <= 40) {
            return Array(3).fill(28);
        }
        if (range <= 80) {
            return Array(3).fill(56);
        }
        if (range <= 160) {
            return Array(3).fill(112);
        }
        return Array(3).fill(224);
    default:
        return Array(3).fill(0);
    }
```

Comment (beheh) [reply src/fmgc/src/efis/EfisCommon.ts:76]: Personally I find the original more readable.

Comment (tracernz) [reply src/fmgc/src/efis/EfisCommon.ts:76]: Particularly as it’s consistent with the other modes which aren’t all the same number.

**Comment by Infrecon** [comment]
Potential bug, DEGES3W departure from LSZH RWY 16. 

Probably not yet implemented Leg Type?

DEGES3S
DEGES3S Chart

File: src/fmgc/src/efis/EfisVectors.ts [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+        const visibleTemporaryFlightPlanVectors = temporaryFlightPlanVectors
+            .filter((vector) => this.vectorWithinCurrentEditArea(vector, 'L') || this.vectorWithinCurrentEditArea(vector, 'R'));
+
+        // ACTIVE
+
+        const engagedLateralMode = SimVar.GetSimVarValue('L:A32NX_FMA_LATERAL_MODE', 'Number') as LateralMode;
+        const armedLateralMode = SimVar.GetSimVarValue('L:A32NX_FMA_LATERAL_ARMED', 'Enum');
+        const navArmed = (armedLateralMode >> 0) & 1;
+
+        const transmitActive = engagedLateralMode === LateralMode.NAV || navArmed;
```
Comment (lukecologne) [inline src/fmgc/src/efis/EfisVectors.ts:60]: Active vectors should be transmitted if any lateral autocontrol mode is active, so if either NAV or LOC is armed, or if NAV, LOC, LOC*, LAND, FLARE, ROLL OUT, RWY, GA TRK or RWY TRK is active
```suggestion
        const navArmed = (armedLateralMode >> 0) & 1;
        const navActive = engagedLateralMode === LateralMode.NAV;
        const takeoffGoaroundModeActive = engagedLateralMode === LateralMode.RWY || engagedLateralMode === LateralMode.RWY_TRACK || engagedLateralMode === LateralMode.GA_TRACK;
        const locArmed = (armedLateralMode >> 1) & 1;
        const locActive = engagedLateralMode >= LateralMode.LOC_CPT && engagedLateralMode <= LateralMode.ROLL_OUT;

        const transmitActive = navActive || takeoffGoaroundModeActive || locActive || navArmed || locArmed;
```

**Comment by SkylarHutter** [comment]
Having issues with the ND showing lines properly. Only shows the path if I am zoomed all the way out. Maybe because the fix is so far?

https://i.imgur.com/8vLlJjp.png
https://i.imgur.com/gb421gE.png


https://i.imgur.com/vU0wFyB.png
https://i.imgur.com/TxiLLRn.png

**Comment by Benjozork** [comment]
Yes, I mistakenly implemened an EIS1 limitation which does not seem to exist anymore.

I'll revert / change soon enough.

**Comment by TVOpium** [comment]
Potential bug : 

DTNH RWY27 LOBNA1D Departure
<a href="https://ibb.co/zrbjfYp">charts</a>
<a href="https://ibb.co/2Ngn1Zb">Sans-titre</a>

LFBO Arrival Initial RNAV RWY32L Transition ADIMO2
<a href="https://ibb.co/XkbGf75">ch</a>
<a href="https://ibb.co/PtXc11S">ch02</a>
<a href="https://imgbb.com/">Sans-titre</a>

**Comment by FinniganH** [comment]
> Yes, I mistakenly implemented an EIS1 limitation which does not seem to exist anymore.
> 
> I'll revert / change soon enough.

Has this been reverted as of yet? I grabbed a new download 30 mins ago, still seeing the same issue:



In this shot, it WAS working on 40 a few miles ago, now it's blank:

**Comment by Benjozork** [comment]
No, I have not reverted this yet;

**Comment by Amorisia** [comment]
Hello,

Here is my test results departing from SAWH runway 25 with USU2B SID with SUNLI transition (using Navigraph AIRAC 2112 rev 2)
Here is the chart:
USU2B 25

A screenshot showing I082 misplaced:
before TO

There was also a curved leg that was not part of the flight plan, morphing as I was flying, so I decided to make a video for better analysis.

This is a private video of the departure available on youtube, with some comments about what I was doing in subtitles and in the description.
[https://www.youtube.com/watch?v=TevpO_OJaN0](url)

Hope this helps and again, good job, unless these few bugs this is another big step and amount of work here !

**Comment by FinniganH** [comment]
Modified original flightplan. Was direct VESDO then YXC, decided to go direct YXC, and this is the result:

image

Seems like it's created an invisible line from the prior fix to YXC instead of present position direct.

**Comment by Benjozork** [comment]
I've changed the culling algorithm to exclude vectors >= 322NM in length instead of using the start / end position. This should remove **some** invisible leg issues.

This will cause issues if you have a very long leg in your flight plan. I've added this as a condition for showing the `MAP PARTLY DISPLAYED` message. This is not Airbus' actual length limit, but is a better solution until we have a better solution

image

**Comment by alexr4339** [comment]
Name: Alex R#6663
Date: 19/12/2021
Tier: 1
Version of this PR: Downloaded PR on the same day, around 12 UTC

Description: 
I did multiple test with this PR and discovered a few bugs. I assume the static altitude waypoint bug is known. Issues with the direct to are known aswell. I had the following procedure:
Load a flight on a runway, insert arrival and departure on Init A, aswell as altitude (sometimes) and flaps aswell as vspeeds on the perf page. Then I inserted the respective departure on the F-PLN page and either in the air or on the ground I inserted an arrival too.

General bugs:
Static altitude waypoint. Altitude waypoint seem to be static and dont move. This can mess a procedure severly up (known)
Direct to issues. Messes up the sequencing, no radius between t-p and direct course to fix, nor drawing of connection between t-p and fix (known)
MCDU sometimes black after loading on runway. Note that this occured in 10 to 20 % of the times when loading the aircraft onto the runway. I did not restart MSFS and it was like loading a flight, then either restart or turn back to menu and load another flight.

Procedural bugs:
LSZR KPT1M
The sequencing and description of this procedure is perfectly accurate on the flight plan page, however the drawing was wrong and the AP tried to follow the wrong drawing
LSZRKPT1M


.
EDXW OLDI1N
The selected procedure took longer to draw than usual (around 10s instead of 1 or 2 usually).
Another issue, after XW100 you shall turn right direct to SLT. Due to the increasing airspeed during climbout the radius constantly updates. However after having initiated the right turn, it doesnt update anymore, which will result that you dont turn directly towards the fix anymore, as your speed increeases and your "overshoot" the procedure
EDXWOLDI1N


LSGG BELU6N
The waypoint GVA8 was drwan only at a selected range of 10. Note: The static alt waypoint issue and how it can potentially mess a procedure up can be seen here aswell

**Comment by Benjozork** [comment]
For `LSZR KPT1M`, issue identified and am thinking of a fix. The problem is that constrained turn direction is not being taken into account.

**Comment by Benjozork** [comment]
As the priority right now is fixing the current issues in the publicly available LNAV, we will be adding manual holding to a future PR. Holds part of procedures will still work.

**Comment by alexr4339** [comment]
Name: Alex R#6663
Date of test: 28/12/2021
Downloaded PR on same day, at 10 UTC
Nav Data: MSFS stock

General:
This PR has matured well and is rock solid. Especially with custom waypoints its capable of nearly everything. I wouldnt want to fly without CFMS15 again and I think due to the fact that it is a drastic improvement to CFMS1 it can be merged to exp soon.

Bugs:
Sometimes legs are not connected or drawn as a right/left turn although they should be connected straight. In one case it affected the guidance (LICR GIMEL1X after LONDA towards RCA). Its not a drastic problem as a DTO solved this issue. But the incorrect drawing is surely not beautiful and might irritate users.


It looked normal after the discontinuity was cleared

This one above was weird as I was coming from north



You can see that LONDA and RCA were not connected and I lost NAV guidance. A simple DTO solved it

I know holdings are not part of this PR for now, but the radius was static. I know I was too fast but even after getting back on course with 200 knots the radius was too small, perhaps due to the fact it calculated the radius with 200kts albeit my TAS was greater due to the altitude.



Here it drew the turn as intended

After having passed RCA11 it flew staright and the right turn dissapeared ( I think the drawing of the turn dissappeared as soon as it targeted RCA11. Interesting to know is that the waypoint after RCA11 is called RCA11 too. A DTO and turning right via heading mode solved this issue.

**Comment by p493z3r0** [comment]
Very good improvement, however i've found an Issue with the Vebit4s Departure out of LSZH. According to the charts it should be a Left turn to cross KLO at R180 and then intercept WIL R052 to BREGO. It is drawn correctly on the ND as well but the AP commands a right turn instead of a left turn. 
image

File: src/fmgc/src/guidance/LnavConfig.ts [hunk 3]
```diff
@@ -37,7 +47,7 @@ export const LnavConfig = {
     /**
      * Whether to draw path debug points and print them out
      */
-    DEBUG_PREDICTED_PATH: false,
+    DEBUG_PREDICTED_PATH: true,
```
Comment (tracernz) [inline src/fmgc/src/guidance/LnavConfig.ts]: Did you intend to commit this one? :D

**Review by aguther — APPROVED** [review]

**Pull request merged**